File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
# 포도주 시식

import sys
input = sys.stdin.readline

n = int(input())
wines = [int(input()) for _ in range(n)]
dp = [0]*n # 포도주의 개수

if n == 0 :
    print(0)
    sys.exit()
elif n == 1: # 와인이 한개면, 무조건 마심
    print(wines[0]) # (1)
elif n == 2: # 와인이 두개여도 무조건 마심
    print(wines[0] + wines[1]) # (1,2)
else:
    dp[0] = wines[0]
    dp[1] = wines[0] + wines[1]
    dp[2] = max(dp[1], wines[1] + wines[2], wines[0] + wines[2]) # (1,2), (2,3), (1,3)

    for i in range(3,n):
        # 3번째 전 + 이전 + 현재 , 2번째 전 + 현재, 바로 이전 값 유지 
        dp[i] += max(dp[i-3] + wines[i-1] + wines[i], dp[i-2] + wines[i], dp[i-1])

    print(dp[-1])

File: 백준/Silver/2156. 포도주 시식/포도주 시식.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const solution = (n, w) => {
  if (n == 1) {
    return w[0];
  } else if (n == 2) {
    return w[0] + w[1];
  } else {
    let ans = Array.from({ length: n }, () => 0);
    ans[0] = w[0];
    ans[1] = w[0] + w[1];
    ans[2] = Math.max(ans[1], w[0] + w[2], w[1] + w[2]);

    for (let i = 3; i < n; i++) {
      ans[i] = Math.max(
        ans[i - 3] + w[i - 1] + w[i],
        w[i] + ans[i - 2],
        ans[i - 1]
      );
    }

    return Math.max(...ans)
  }
};

const n = Number(input[0]); // 포도주 잔의 개수
let w = Array.from({ length: n }, () => 0);
for (let i = 1; i <= n; i++) {
  w[i - 1] = Number(input[i].trim());
}
if (n == 0) {
  console.log(0)
  
} else {
  
  console.log(solution(n, w))
}
